### PR TITLE
Finish layout and routes

### PR DESCRIPTION
**Pull request by Phamduc8797**

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
-
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-  </head>
-
-  <body>
-    <%= yield %>
-  </body>
+<head>
+  <title><%= full_title(yield(:title)) %></title>
```
Comment (huuphuc93) [inline app/views/layouts/application.html.erb]: full_title(yield(:title)) => full_title yield(:title)

File: app/assets/javascripts/custom.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// Any JavaScript/Coffee file within this directory, lib/assets/javascripts, or any plugin's
+// vendor/assets/javascripts directory can be referenced here using a relative path.
+//
+// It's not advisable to add code directly here, but if you do, it'll appear at the bottom of the
+// compiled file. JavaScript code in this file should be added after the last require_* statement.
+//
+// Read Sprockets README (https://github.com/rails/sprockets#sprockets-directives) for details
+// about supported directives.
+//
+//= require rails-ujs
```
Comment (huuphuc93) [inline app/assets/javascripts/custom.js]: remove all comment before this line

File: .idea/sample_app.iml [hunk 3]
```diff
@@ -38,12 +38,14 @@
     <orderEntry type="library" scope="PROVIDED" name="ansi (v1.5.0, RVM: ruby-2.5.1) [gem]" level="application" />
```
Comment (huuphuc93) [inline .idea/sample_app.iml]: use git rm --cached file  to remove .idea file from cached before push pull request

File: app/assets/javascripts/custom.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+//= require activestorage
+//= require turbolinks
+//= require_tree .
```
Comment (huuphuc93) [inline app/assets/javascripts/custom.js]: add break line after the end of file

File: app/views/layouts/_footer.html.erb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<footer class="footer">
+  <small>
+    The <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
```
Comment (huuphuc93) [inline app/views/layouts/_footer.html.erb]: Use I18n https://github.com/svenfuchs/i18n

File: app/views/static_pages/home.html.erb [hunk 6]
```diff
@@ -1,14 +1,10 @@
...
+<div class="center jumbotron">
+  <h1>Welcome to the Sample App</h1>
+  <h2>
+    This is the home page for the
+    <a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>
+    sample application.
+  </h2>
+  <%= link_to "Sign up now!", "/signup", class: "btn btn-lg btn-primary" %>
+</div>
+<%= link_to image_tag("rails.png", alt: "Rails logo"), "http://rubyonrails.org/" %>
```
Comment (huuphuc93) [inline app/views/static_pages/home.html.erb]: remove ()

File: Gemfile [hunk 7]
```diff
@@ -2,7 +2,8 @@ source 'https://rubygems.org'
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
 ruby '2.5.1'
-
+gem "sprockets-rails", :require => "sprockets/railtie"
+gem "bootstrap-sass", "~> 3.3.6"
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails'
```
Comment (huuphuc93) [inline Gemfile]: remove all comments, use " instead of ' in this whole file

File: app/assets/stylesheets/application.css [hunk 8]
```diff
@@ -1,15 +0,0 @@
...
- * vendor/assets/stylesheets directory can be referenced here using a relative path.
- *
- * You're free to add application-wide styles to this file and they'll appear at the bottom of the
- * compiled file so the styles you add here take precedence over styles defined in any other CSS/SCSS
- * files in this directory. Styles in this file should be added after the last require_* statement.
- * It is generally better to create a new file per style scope.
- *
- *= require_tree .
- *= require_self
- */
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/application.css]: Why you have removed this. Please restore this file again.

File: app/assets/stylesheets/custom.scss [hunk 9]
```diff
@@ -0,0 +1,102 @@
+@import 'bootstrap-sprockets';
+@import 'bootstrap';
+$color_w: #fff;
+$color_g1: #222;
+$color_g2: #777;
+$color_g3: #555;
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Add a line break here

File: app/assets/stylesheets/custom.scss [hunk 10]
```diff
@@ -0,0 +1,102 @@
+@import 'bootstrap-sprockets';
+@import 'bootstrap';
```
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Add a line break here

File: app/assets/stylesheets/custom.scss (hunk 9 above)
Comment (mdmahmudurrahman) [inline app/assets/stylesheets/custom.scss]: Move those color variable to an external file(example color_variable.scss), then export them from there.

File: app/models/user.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+	before_save { self.email = email.downcase }
+	validate :name, presence: true, length: { maximum: 50 }
+	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+	validate :email, presence: true, length: { maximum: 255 },
+					format: { with: VALID_EMAIL_REGEX },
+					uniqueness: { case_sensitive: false } #xac dinh tinh trung lap
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Indentation

File: app/views/layouts/_footer.html.erb [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<footer class="footer">
+  <small>
+    <%= t ".the" %> <a href="http://www.railstutorial.org/"><%= t ".railstutorial" %></a>
+    <%= t ".by" %> <a href="http://www.michaelhartl.com/"><%= t ".michaelhartl" %></a>
+  </small>
+  <nav>
+    <ul>
+      <li><%= link_to "About", "/about" %></li>
+      <li><%= link_to "Contact", "/contact" %></li>
+      <li><a href="http://news.railstutorial.org/">News</a></li>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/_footer.html.erb]: Use I18n for plain text
Reference [https://guides.rubyonrails.org/i18n.html](https://guides.rubyonrails.org/i18n.html)

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -1,15 +1,17 @@
...
-  </body>
+<head>
+  <title><%= full_title yield(:title) %></title>
+  <%= csrf_meta_tags %>
+  <%= stylesheet_link_tag "custom", media: "all",
+                             "data-turbolinks-track": "reload" %>
+  <%= javascript_include_tag "custom", "data-turbolinks-track": "reload" %>
+</head>
+<body>
+<%= render "layouts/header" %>
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: Indentation

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -1,15 +1,17 @@
...
+<div class="container">
```
Comment (mdmahmudurrahman) [inline app/views/layouts/application.html.erb]: Same here

File: app/views/static_pages/about.html.erb [hunk 15]
```diff
@@ -1,18 +1,17 @@
-<% provide(:title, "About") %>
+<% provide :title, "About" %>
```
Comment (mdmahmudurrahman) [inline app/views/static_pages/about.html.erb]: Implement I18n

File: app/views/users/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,3 @@
+<% provide :title, "Sign up" %>
+<h1>Sign up</h1>
+<p>This will be a signup page for new users.</p>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<% provide :title, "Sign up" %>
+<h1>Sign up</h1>
```
Comment (mdmahmudurrahman) [inline app/views/users/new.html.erb]: I18n

File: config/locales/en.yml [hunk 18]
```diff
@@ -1,33 +1,34 @@
...
+      firsthelp: "Get help on the Ruby on Rails Tutorial at the"
+      railshelp: "Rails Tutorial help section"
+      nexthelp: "To get help on this sample app, see the"
+      railstutorial: "Ruby on Rails Tutorial"
+      book: "book"
+    home:
+      welcome: "Welcome to the Sample App"
+      firstwel: "This is the home page for the"
+      railstutorial: "Ruby on Rails Tutorial"
+      lastwel: "sample application."
```
Comment (mdmahmudurrahman) [inline config/locales/en.yml]: Add a line break at the end of a file

File: app/models/user.rb [hunk 19]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+	VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (huuphuc93) [inline app/models/user.rb]: ident 2 spaces instead of 4

File: app/models/user.rb [hunk 20]
```diff
@@ -0,0 +1,6 @@
...
+	before_save { self.email = email.downcase }
```
Comment (huuphuc93) [inline app/models/user.rb]: { self.email = email.downcase } => {self.email = email.downcase}

File: app/views/users/new.html.erb [hunk 21]
```diff
@@ -0,0 +1,3 @@
+<% provide :title, "Sign up" %>
+<h1><%= t ".signup" %></h1>
+<p><%= t ".content" %>
```
Comment (huuphuc93) [inline app/views/users/new.html.erb]: add break line at the end of this file

**Comment by Phamduc8797** [comment]
ready

**Comment by Phamduc8797** [comment]
ready

**Comment by Phamduc8797** [comment]
ready

**Comment by Phamduc8797** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 22]
```diff
@@ -0,0 +1,102 @@
+@import 'bootstrap-sprockets';
+@import 'bootstrap'; 
+
```
Comment (huuphuc93) [inline app/assets/stylesheets/custom.scss]: remove this empty line

File: app/models/user.rb [hunk 23]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+	before_save {self.email = email.downcase}
```
Comment (huuphuc93) [inline app/models/user.rb]: ident this line and 2 line after

File: app/views/layouts/_footer.html.erb [hunk 24]
```diff
@@ -0,0 +1,13 @@
+<footer class="footer">
+  <small>
+    <%= t ".the" %> <a href="http://www.railstutorial.org/"><%= t ".railstutorial" %></a>
```
Comment (huuphuc93) [inline app/views/layouts/_footer.html.erb]: move "http://www.railstutorial.org/" to file Settings

**Comment by huuphuc93** [comment]
commented 
you need fix all comments before push pull

**Comment by Phamduc8797** [comment]
ready

File: app/models/user.rb [hunk 25]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = Setting.VALID_EMAIL_REGEX
+	before_save {self.email = email.downcase}
```
Comment (huuphuc93) [inline app/models/user.rb]: ident

File: app/views/layouts/application.html.erb [hunk 26]
```diff
@@ -1,15 +1,16 @@
...
-    <%= csp_meta_tag %>
-
-    <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-  </head>
-
-  <body>
-    <%= yield %>
-  </body>
+<head>
```
Comment (huuphuc93) [inline app/views/layouts/application.html.erb]: indent

File: app/views/static_pages/help.html.erb [hunk 27]
```diff
@@ -1,17 +1,16 @@
-<% provide(:title, "Help") %>
+<% provide :title, t(".help") %>
 <!DOCTYPE html>
```
Comment (huuphuc93) [inline app/views/static_pages/help.html.erb]: why this file have <!DOCTYPE html> tag?

File: app/views/static_pages/home.html.erb [hunk 28]
```diff
@@ -1,14 +1,11 @@
...
+  <h1><%= t ".welcome" %></h1>
+  <h2>
+    <%= t ".firstwel" %>
+    <a href=<%= Settings.railstutorial %> ><%= t ".railstutorial" %></a>
+    <%= t ".lastwel" %>
+  </h2>
+  <%= link_to t(".signupnow"), "/signup", class: "btn btn-lg btn-primary" %>
+</div>
+<%= link_to image_tag("rails.png", alt: "Rails logo"),
+<%= Settings.railstutorial %> %>
```
Comment (huuphuc93) [inline app/views/static_pages/home.html.erb]: đoạn này chạy dc ko  em?

**Comment by huuphuc93** [comment]
commented

**Comment by Phamduc8797** [comment]
ready

**Comment by Phamduc8797** [comment]
ready

File: app/views/static_pages/about.html.erb [hunk 29]
```diff
@@ -1,18 +1,11 @@
...
-  The <a href="http://www.railstutorial.org/"><em>Ruby on Rails
-  Tutorial</em></a> is a
-  <a href="http://www.railstutorial.org/book">book</a> and
-  <a href="http://screencasts.railstutorial.org/">screencast series</a>
-  to teach web development with
-  <a href="http://rubyonrails.org/">Ruby on Rails</a>.
-  This is the sample application for the tutorial.
+  <%= t ".the" %> <a href=<%= Settings.railstutorial %> ><em>
+  <%= t ".railstutorial" %></em></a> <%= t ".isa" %>
+  <a href=<%= Settings.railsbook %> ><%= t ".book" %></a> <%= t ".and" %>
```
Comment (huuphuc93) [inline app/views/static_pages/about.html.erb]: use <%= link_to %> intead of <a></a>

File: app/views/static_pages/contact.html.erb [hunk 30]
```diff
@@ -1,6 +1,6 @@
...
-<h1>Contact</h1>
+<% provide :title, t(".title")%>
+<h1><%= t ".contact" %></h1>
 <p>
-  Contact the Ruby on Rails Tutorial about the sample app at the
-  <a href="http://www.railstutorial.org/contact">contact page</a>.
-</p>
\ No newline at end of file
+  <%= t ".first" %>
+  <a href=<%= Settings.railscontact %> ><%= t ".contactpage" %></a>.
```
Comment (huuphuc93) [inline app/views/static_pages/contact.html.erb]: use <%= link_to %> intead of

File: app/views/static_pages/help.html.erb [hunk 31]
```diff
@@ -1,17 +1,9 @@
...
+<h1><%= t ".help" %></h1>
 <p>
-  Get help on the Ruby on Rails Tutorial at the
-  <a href="http://www.railstutorial.org/help">Rails Tutorial help
-    section</a>.
-  To get help on this sample app, see the
-  <a href="http://www.railstutorial.org/book"><em>Ruby on Rails
-    Tutorial</em> book</a>.
+  <%= t ".firsthelp" %>
+  <a href=<%= Settings.railshelp %> ><%= t ".railshelp" %></a>.
```
Comment (huuphuc93) [inline app/views/static_pages/help.html.erb]: use <%= link_to %> intead of

File: app/views/static_pages/home.html.erb [hunk 32]
```diff
@@ -1,14 +1,11 @@
...
-  sample application.
-</p>
-</body>
-</html>
\ No newline at end of file
+<div class="center jumbotron">
+  <h1><%= t ".welcome" %></h1>
+  <h2>
+    <%= t ".firstwel" %>
+    <a href=<%= Settings.railstutorial %> ><%= t ".railstutorial" %></a>
```
Comment (huuphuc93) [inline app/views/static_pages/home.html.erb]: use <%= link_to %> intead of

**Comment by huuphuc93** [comment]
commented

**Comment by Phamduc8797** [comment]
ready

**Pull request merged**